python/ray/train/torch.py

```diff
@@ -98,34 +97,38 @@ def prepare_model(
             logger.info(f"Moving model to device: {device}")
             model = model.to(device)
 
-        def wrap_forward(forward):
-            @functools.wraps(forward)
-            def wrapper(*args, **kwargs):
-                with autocast():
-                    outputs = forward(*args, **kwargs)
-                assert isinstance(outputs, torch.Tensor)
-                return outputs.float()
-
-            return wrapper
-
         def model_get_state(self):
             # `__getstate__` is an special method that informs pickle which attributes
             # to serialize. This custom implementation ensures that the wrapped forward
             # method and custom `__getstate__` method aren't serialized.
-            state = self.__dict__.copy()
+            if hasattr(self, "_original_get_state"):
+                state = self._original_get_state()
+                state["__getstate__"] = state["_original_get_state"]
+                del state["_original_get_state"]
+            else:
+                # If model does not have a __getstate__ already defined, use default
+                # implementation.
+                state = self.__dict__.copy()
+                del state["__getstate__"]
             state["forward"] = state["_unwrapped_forward"]
             del state["_unwrapped_forward"]
-            del state["__getstate__"]
+
             return state
 
         if self.amp_is_enabled:
             # Pickle cannot serialize the wrapped forward method. As a workaround,
             # define a custom `__getstate__` method that unwraps the forward method.
             model._unwrapped_forward = model.forward
-            model.forward = wrap_forward(model.forward)
+            model.forward = autocast()(model.forward)
```

Comment:
Question - why is `wrap_forward` no longer needed here?

Reply:
The new change is equivalent and we don't have to define our own function. `autocast` provides a function interface already so we don't have to do it ourselves.

```
    def __call__(self, func):
        @functools.wraps(func)
        def decorate_autocast(*args, **kwargs):
            with self:
                return func(*args, **kwargs)
        return decorate_autocast
```